src/OrleansCodeGenerator/RoslynCodeGenerator.cs

```diff
@@ -426,12 +426,18 @@ private static GeneratedSyntax GenerateForAssemblies(List<Assembly> assemblies,
             var typeInfo = type.GetTypeInfo();
 
             // Every type which is encountered must be considered for serialization.
-            if (!typeInfo.IsNested && !typeInfo.IsGenericParameter && typeInfo.IsSerializable)
+            if (!typeInfo.IsNested && !typeInfo.IsGenericParameter)
             {
-                // If a type was encountered which can be accessed, process it for serialization.
-                if (SerializerGenerationManager.RecordTypeToGenerate(type, module, targetAssembly))
+                // If a type was encountered which can be accessed and is marked as [Serializable], process it for serialization.
+                if (typeInfo.IsSerializable)
+                    RecordType(type, module, targetAssembly, includedTypes);
+
+                Type grainStateType;
+                // If a type extends Grain<T>, add T to the list of candidates for serialization
+                if (IsPersistentGrain(typeInfo, out grainStateType))
                 {
-                    includedTypes.Add(type);
+                    if (!grainStateType.IsNested && !grainStateType.IsGenericParameter)
```

Comment:
We could probably remove the `.IsNested` requirement and shift the `IsGenericParameter` requirement into `SerializerGeneratorManager.RecordTypeToGenerate` if it isn't there already (it should be in there)